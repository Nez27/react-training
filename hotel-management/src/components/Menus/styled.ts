import styled from 'styled-components';

const StyledMenu = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;

const StyledButton = styled.button`
  width: 100%;
  text-align: left;
  background: none;
  border: none;
  padding: 10px 20px;
  font-size: var(--fs-sm-x);
  transition: all 0.2s;

  display: flex;
  align-items: center;
  gap: 10px;

  &:hover {
    background-color: var(--hover-background-color);
  }

  & svg {
    width: 18px;
    height: 18px;
    color: var(--primary-color);
    transition: all 0.3s;
  }
`;

const StyledToggle = styled.button`
  background: none;
  border: none;
  padding: 8px;
  border-radius: var(--radius-sm);
  transform: translateX(5px);
  transition: all 0.2s;

  &:hover {
    background-color: var(--hover-background-color);
  }

  & svg {
    width: 18px;
    height: 18px;
  }
`;

const StyledList = styled.ul`
  position: absolute;
  z-index: 1000;

  background-color: white;
  box-shadow: var(--shadow-sm);
  border-radius: var(--radius-sm);

  right: -10px;
  top: 40px;
`;

export { StyledMenu, StyledButton, StyledList, StyledToggle };
