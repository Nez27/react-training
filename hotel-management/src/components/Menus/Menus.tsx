import { useContext, useState } from 'react';

// Components
import { HiEllipsisVertical } from 'react-icons/hi2';
import { useOutsideClick } from '../../hooks/useOutsideClick';
import { StyledMenu, StyledButton, StyledList, StyledToggle } from './styled';

// Interfaces
import { IButton } from '../../globals/interfaces';

// Contexts
import MenusContext from '../../contexts/MenuContext';

const Menus = ({ children }: { children: JSX.Element }) => {
  const [openId, setOpenId] = useState('');

  const close = () => setOpenId('');
  const open = setOpenId;

  return (
    <MenusContext.Provider value={{ openId, close, open }}>
      {children}
    </MenusContext.Provider>
  );
};

const Toggle = ({ id }: { id: string }): React.JSX.Element => {
  const { openId, close, open } = useContext(MenusContext);

  const handleClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.stopPropagation();

    // prettier-ignore
    openId === '' || openId !== id 
      ? open!(id) 
      : close!();
  };

  return (
    <StyledToggle onClick={handleClick}>
      <HiEllipsisVertical />
    </StyledToggle>
  );
};

const List = ({
  id,
  children,
}: {
  id: string;
  children: JSX.Element[];
}): React.JSX.Element | null => {
  const { openId, close } = useContext(MenusContext);
  const ref = useOutsideClick(close!, false);

  if (openId !== id) return null;

  return <StyledList ref={ref}>{children}</StyledList>;
};

const Button = ({ children, icon, onClick }: IButton): React.JSX.Element => {
  const { close } = useContext(MenusContext);

  const handleClick = () => {
    onClick?.();
    close!();
  };

  return (
    <li>
      <StyledButton onClick={handleClick}>
        {icon} <span>{children}</span>
      </StyledButton>
    </li>
  );
};

Menus.Menu = StyledMenu;
Menus.Toggle = Toggle;
Menus.List = List;
Menus.Button = Button;

export default Menus;
