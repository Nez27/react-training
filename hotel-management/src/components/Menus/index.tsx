import { MouseEvent, ReactNode, useContext, useState } from 'react';

// Components
import { HiEllipsisVertical } from 'react-icons/hi2';
import ButtonIcon from '../ButtonIcon';

// Hooks
import { useOutsideClick } from '@src/hooks/useOutsideClick';

// Styled
import { StyledMenu, StyledList, StyledToggle } from './styled';

// Contexts
import MenusContext from '@src/contexts/MenuContext';

// Types
import { Nullable } from '@src/types/common';

interface IButton {
  label?: string;
  icon?: ReactNode;
  onClick?: () => void;
  disabled?: boolean;
}

const Menus = ({ children }: { children: ReactNode }) => {
  const [openId, setOpenId] = useState('');
  const close = () => setOpenId('');
  const open = setOpenId;

  return (
    <MenusContext.Provider
      value={{
        openId,
        close,
        open,
      }}
    >
      {children}
    </MenusContext.Provider>
  );
};

const Toggle = ({ id }: { id: string }): ReactNode => {
  const { openId, close, open } = useContext(MenusContext);
  const handleClick = (e: MouseEvent<HTMLButtonElement>) => {
    e.stopPropagation();

    openId !== id ? open!(id) : close!();
  };

  return (
    <StyledToggle onClick={handleClick} aria-label={`Menu item ${id}`}>
      <HiEllipsisVertical />
    </StyledToggle>
  );
};

const List = ({
  id,
  children,
}: {
  id: string;
  children: ReactNode;
}): Nullable<ReactNode> => {
  const { openId, close } = useContext(MenusContext);
  const ref = useOutsideClick<HTMLUListElement>(close!, false);

  if (openId !== id) return null;

  return <StyledList ref={ref}>{children}</StyledList>;
};

const Button = ({ label, icon, onClick, disabled }: IButton): ReactNode => {
  const { close } = useContext(MenusContext);
  const handleClick = () => {
    onClick?.();
    close!();
  };

  return (
    <li>
      <ButtonIcon
        text={label}
        aria-label={label}
        icon={icon}
        iconColor='var(--primary-color)'
        onClick={handleClick}
        disabled={disabled}
      />
    </li>
  );
};

Menus.Menu = StyledMenu;
Menus.Toggle = Toggle;
Menus.List = List;
Menus.Button = Button;

export default Menus;
