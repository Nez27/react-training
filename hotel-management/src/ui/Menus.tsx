import { createContext, useContext, useEffect, useRef, useState } from 'react';
import { HiEllipsisVertical } from 'react-icons/hi2';
import styled from 'styled-components';

const StyledMenu = styled.div`
  position: relative;
  display: flex;
  align-items: center;
  justify-content: flex-end;
`;

const StyledButton = styled.button`
  width: 100%;
  text-align: left;
  background: none;
  border: none;
  padding: 10px 20px;
  font-size: var(--fs-sm-x);
  transition: all 0.2s;

  display: flex;
  align-items: center;
  gap: 10px;

  &:hover {
    background-color: var(--hover-background-color);
  }

  & svg {
    width: 18px;
    height: 18px;
    color: var(--primary-color);
    transition: all 0.3s;
  }
`;

const StyledToggle = styled.button`
  background: none;
  border: none;
  padding: 8px;
  border-radius: var(--radius-sm);
  transform: translateX(5px);
  transition: all 0.2s;

  &:hover {
    background-color: var(--hover-background-color);
  }

  & svg {
    width: 18px;
    height: 18px;
  }
`;

const StyledList = styled.ul`
  position: absolute;
  z-index: 1000;

  background-color: white;
  box-shadow: var(--shadow-sm);
  border-radius: var(--radius-sm);

  right: -10px;
  top: 40px;
`;

// ------------- Interface ------------- //
interface IMenusContext {
  openId?: string;
  close?: () => void;
  open?: React.Dispatch<React.SetStateAction<string>>;
}

interface IButton {
  children?: string;
  icon?: JSX.Element;
  onClick?: () => void;
}

// ------------- Custom hooks ------------- //
const useOutsideClick = (
  handler: () => void,
  listeningCapturing = true,
): React.MutableRefObject<HTMLUListElement | null> => {
  const ref = useRef<HTMLUListElement>(null);

  useEffect(() => {
    const handleClick = (e: Event) => {
      if (ref.current && !ref.current.contains(e.target as Node)) {
        handler();
      }
    };

    document.addEventListener('click', handleClick, listeningCapturing);

    return () =>
      document.removeEventListener('click', handleClick, listeningCapturing);
  }, [handler, listeningCapturing]);

  return ref;
};

// ------------- Menu Component ------------- //
const MenusContext = createContext<IMenusContext>({});

const Menus = ({ children }: { children: JSX.Element }) => {
  const [openId, setOpenId] = useState('');

  const close = () => setOpenId('');
  const open = setOpenId;

  return (
    <MenusContext.Provider value={{ openId, close, open }}>
      {children}
    </MenusContext.Provider>
  );
};

const Toggle = ({ id }: { id: string }): React.JSX.Element => {
  const { openId, close, open } = useContext(MenusContext);

  const handleClick = (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => {
    e.stopPropagation();

    openId === '' || openId !== id ? open!(id) : close!();
  };

  return (
    <StyledToggle onClick={(e) => handleClick(e)}>
      <HiEllipsisVertical />
    </StyledToggle>
  );
};

const List = ({
  id,
  children,
}: {
  id: string;
  children: JSX.Element[];
}): React.JSX.Element | null => {
  const { openId, close } = useContext(MenusContext);
  const ref = useOutsideClick(close!, false);

  if (openId !== id) return null;

  return <StyledList ref={ref}>{children}</StyledList>;
};

const Button = ({ children, icon, onClick }: IButton): React.JSX.Element => {
  const { close } = useContext(MenusContext);

  const handleClick = () => {
    onClick?.();
    close!();
  };

  return (
    <li>
      <StyledButton onClick={handleClick}>
        {icon} <span>{children}</span>
      </StyledButton>
    </li>
  );
};

Menus.Menu = StyledMenu;
Menus.Toggle = Toggle;
Menus.List = List;
Menus.Button = Button;

export default Menus;
