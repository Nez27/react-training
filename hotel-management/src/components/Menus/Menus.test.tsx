import renderer from 'react-test-renderer';

// Components
import Menus from '.';
import { RiEditBoxFill, RiDeleteBin2Line } from 'react-icons/ri';

describe('Menus snapshot testing', () => {
  const id = '123';
  const handleOnClick = () => {
    console.log('Click');
  };

  const wrapper = renderer.create(
    <Menus>
      <Menus.Menu>
        <Menus.Toggle id={id} />

        <Menus.List id={id.toString()}>
          <Menus.Button onClick={handleOnClick} icon={<RiEditBoxFill />}>
            Edit
          </Menus.Button>
          <Menus.Button onClick={handleOnClick} icon={<RiDeleteBin2Line />}>
            Delete
          </Menus.Button>
        </Menus.List>
      </Menus.Menu>
    </Menus>
  );

  test('render', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
