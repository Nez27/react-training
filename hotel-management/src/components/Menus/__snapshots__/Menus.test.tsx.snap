// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menus snapshot testing render 1`] = `
<div
  className="sc-bdnyFh cumhtC"
>
  <button
    aria-label="Menu item 123"
    className="sc-dlniIP jergss"
    onClick={[Function]}
  >
    <svg
      aria-hidden="true"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        {
          "color": undefined,
        }
      }
      viewBox="0 0 24 24"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        clipRule="evenodd"
        d="M10.5 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0zm0 6a1.5 1.5 0 113 0 1.5 1.5 0 01-3 0z"
        fillRule="evenodd"
      />
    </svg>
  </button>
</div>
`;
